File: app/src/main/java/org/wikipedia/analytics/FunnelManager.java
package org.wikipedia.analytics;

import org.wikipedia.WikipediaApp;
import org.wikipedia.page.PageTitle;

import java.util.Hashtable;
import java.util.Map;

/**
 * Creates and stores analytics tracking funnels.
 */
public class FunnelManager {
    private final WikipediaApp app;
    private final Map<PageTitle, EditFunnel> editFunnels = new Hashtable<>();

    public FunnelManager(WikipediaApp app) {
        this.app = app;
    }

    public EditFunnel getEditFunnel(PageTitle title) {
        return editFunnels.computeIfAbsent(title, key -> new EditFunnel(app, key));
    }
}

File: app/src/main/java/org/wikipedia/analytics/FunnelManager.kt
package org.wikipedia.analytics

import org.wikipedia.WikipediaApp
import org.wikipedia.page.PageTitle
import java.util.concurrent.ConcurrentHashMap

class FunnelManager(private val app: WikipediaApp) {
    private val editFunnels = ConcurrentHashMap<PageTitle, EditFunnel>()

    fun getEditFunnel(title: PageTitle): EditFunnel {
        return editFunnels.getOrPut(title) { EditFunnel(app, title) }
    }
}
